File: java-algorithms/src/main/java/com/jcdevbot/datastruct/binarytree/BinaryTree.java
package com.jcdevbot.datastruct.binarytree;

import java.util.ArrayList;

/**
 * Created by jcheung on 6/10/16.
 */
public class BinaryTree{
    private Node root;

    public BinaryTree(Node root){
        this.root = root;
    }

    public void insert(Node  node){
        if(root == null){
            root = node;
        }else{
            insertNode(root,node);
        }

    }

    private void insertNode(Node node, Node newNode){
        Node left = node.getLeft();
        Node right = node.getRight();

        if(node.getData() > newNode.getData()){
            if(left != null){
                insertNode(left,newNode);
            }else{
                node.setLeft(newNode);
            }
        }else{
            if(right != null){
                insertNode(right,newNode);
            }else{
                node.setRight(newNode);
            }
        }
    }

    public void printTree(){
        inOrderTraversal(root);
    }

    private void postOrderTraversal(Node node){
        System.out.println(node.getData());
    }

    private void preOrderTraversal(Node node){
        System.out.println(node.getData());

    }
    private void inOrderTraversal(Node node){
        if(node.getLeft() != null){
            inOrderTraversal(node.getLeft());
        }
        System.out.println(node.getData());
        if(node.getRight() != null){
            inOrderTraversal(node.getRight());
        }
    }

    private void convertTreeToArray(Node node, ArrayList list){
        if(node.getLeft() != null){
            convertTreeToArray(node.getLeft(),list);
        }
        list.add(node);
        if(node.getRight() != null){
            convertTreeToArray(node.getRight(),list);
        }
    }

    public ArrayList getBinaryTreeAsArray(){
        ArrayList<Node> list = new ArrayList<Node>();
        convertTreeToArray(root,list);

        for(Node node : list){
            System.out.print(node.getData() + " ");
        }
        return list;
    }

    public int getMaxDepth(){
        return getMaxDepth(root);
    }

    private int getMaxDepth(Node node){
        if(node == null) return 0;
        int left_height = getMaxDepth(node.getLeft());
        int right_height = getMaxDepth(node.getRight());
        return (left_height > right_height) ? left_height + 1 : right_height + 1;
    }

}

File: myAlgorithmns/binary_tree.cpp
#include <iostream> 
using namespace std;

struct Node{

	int data;
	Node *left;
	Node *right;

} node;


void printTree(Node *root){

	cout << root->data << endl;

	if(root->left != NULL){
		printTree(root->left);
	}

	if(root->right != NULL){
		printTree(root->right);
	}
}

class btree
{
	public:
		btree();
		~btree();
		void insert(int key);
		Node *search(int key);
		void destroy_tree();
		void print_tree();
	private:
		void insert(int key, Node *leaf);
		Node *search(int key, Node *leaf);
		Node *root;

};

btree::btree(){
	root = NULL;
}

void btree::insert(int key, Node *leaf){
	if(key < leaf->data){
		if(leaf->left != NULL)
			insert(key, leaf->left);
		else{
			leaf->left = new Node;
			leaf->left->data = key;
			leaf->left->left = NULL;
			leaf->left->right = NULL;
		}
	}else if(key >= leaf->data){
		if(leaf->right != NULL)
			insert(key, leaf->right);
		else{
			leaf->right = new Node;
			leaf->right->data = key;
			leaf->right->left = NULL;
			leaf->right->right = NULL;
		}
	}
}

Node *btree::search(int key,Node *leaf){
	if(leaf != NULL){
		if(key==leaf->data){
			return leaf;
		}else if(key < leaf->data){
			return search(key,leaf->left);
		}else{
			return search(key,leaf->right);
		}
	}

}

void btree::insert(int key){
	if(root == NULL){
		root = new Node;
		root->data = key;
		root->left = NULL;
		root->right = NULL;
	}else{
		insert(key,root);
	}
}

void btree::print_tree(){

}

btree::~btree(){

}


int main(){

	btree tree;
	tree.insert(1);
	tree.insert(2);
	tree.insert(3);
}
